### PR TITLE
Fix slow call to scandir() when uploading files

### LINKED ISSUE #373: Extremely slow _wp_handle_upload()
Hey guys,

Love this plugin - been using it for years.

I've stumbled across a problem with `_wp_handle_upload()` where it's taking way too long to handle the uploaded file. I discovered this when trying to import attachments programmatically with `media_handle_sideload()`.

I've wrote a small script on a fresh WP install to test this out with an 12KB image.

```php
$starttime_upload = microtime(true);
_wp_handle_upload($file, ['test_form' => false, 'test_size' => false], current_time('mysql'), 'wp_handle_sideload');
$endtime_upload = microtime(true) - $starttime_upload;

echo "took {$endtime_upload}s to handle upload";
```

Which outputs:
```
took 2.6300148963928s to handle upload
```

If I disable S3-Uploads plugin I get:
```
took 0.029120206832886s to handle upload
```

Just to check I ran `wp s3-uploads verify` and it verified in around 2-300ms.

Not sure if this is expected or not - would appreciate any input you guys might have on this.

### PR DESCRIPTION
**Pull request by joehoyle**
Please for give me for what I have done.

Fixes #373.

### CODE REVIEW

**Comment by joehoyle** [comment]
@rmccue could I get a 👍 / 👎  today on this, as this is currently blocking some production errors.

**Comment by joehoyle** [comment]
@rmccue I added support for the filter too here.

File: inc/class-stream-wrapper.php [hunk 1]
```diff
@@ -583,11 +583,45 @@ public function dir_opendir( $path, $options ) {
...
+		// to include the filename, therefore only return a subset of files from S3 that are more likely to match
+		// the preg_match() call.
+		//
+		// Essentially, wp_unique_filename( my-file.jpg ) doing a `scandir( s3://bucket/2019/04/ )` will actually result in an s3
+		// listObject query for `s3://bucket/2019/04/my-file` which means even if there are millions of files in `2019/04/` we only
+		// return a much smaller subset.
+		//
+		// Anyone reading this far, brace yourselves for a mighty horrible hack.
+		$backtrace = debug_backtrace( 0, 3 ); // phpcs:ignore PHPCompatibility.FunctionUse.ArgumentFunctionsReportCurrentValue.NeedsInspection
+		if ( isset( $backtrace[1]['function'] ) && $backtrace[1]['function'] === 'scandir' && isset( $backtrace[2]['function'] ) && $backtrace[2]['function'] === 'wp_unique_filename' ) {
```
Comment (rmccue) [inline inc/class-stream-wrapper.php:620]: Does this need to bail if it hit the filter callback above, or is that already handled?

Comment (joehoyle) [reply inc/class-stream-wrapper.php:620]: This won't be triggered, as the backtrace doesn't reflect our own `scandir` call.

Comment (rmccue) [reply inc/class-stream-wrapper.php:620]: Gotcha.

**Pull request merged**